src/cargo/util/toml/mod.rs

```diff
@@ -1922,7 +1959,21 @@ impl TomlManifest {
                     .clone()
                     .resolve_with_self(n, |dep| dep.resolve(n, inheritable, cx))?;
                 let dep = resolved.to_dependency(n, cx, kind)?;
-                validate_package_name(dep.name_in_toml().as_str(), "dependency name", "")?;
+                let name_in_toml = dep.name_in_toml().as_str();
+                validate_package_name(name_in_toml, "dependency name", "")?;
+                let kind_name = &kind
+                    .map(|k| match k {
+                        DepKind::Normal => "dependencies",
+                        DepKind::Development => "dev-dependencies",
+                        DepKind::Build => "build-dependencies",
+                    })
+                    .unwrap_or("dependencies");
+                let table_in_toml = if let Some(platform) = &cx.platform {
+                    format!("target.{}.{kind_name}", platform.to_string())
```

Comment:
Note: This doesn't handle quoting TOML keys. But it looks like `stringify` doesn't do that, either. Probably not too important, just wanted to note it.